clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -317,6 +342,16 @@ public void reenqueue(ProducerBatch batch, long now) {
         }
     }
 
+    private void maybeRemoveFromInflight(ProducerBatch batch) {
```

Comment:
Shouldn't we also be removing the batches from the inflight set when the batch is completed (failed or successfully)? I might be missing something, but I don't see that code here.

Reply:
@apurvam Right. Cleanup of `soonToExpireInFlightBatches` happens in two places (1) if a batch gets reenqueued and (2) when `Sender` looks for `expiredBatches`. In the second case, we cleanup the batches whose final state is known (success or failure) and there by "removing" them.

Reply:
Hmm. This seems a bit off. What this means that in the 'normal' case when responses are successful and there is no backlog in the accumulator, we will hang on to batches (and not garbage collect them) until the delivery timeout. 

Indeed, if you add the following at the end of sender tests where there are no more inflight requests (ie. all requests have completed and will never be retried) `        assertTrue(accumulator.soonToExpireInFlightBatches().isEmpty());` all the tests fail.

I think this should be fixed. We should clear the `soonToExpireInflightBatches` as soon as the batch is completely resolved (ie. failed or completed) so that we don't hang on to the reference unnecessarily.